pages/v4/Coroutines.md

```diff
@@ -0,0 +1,184 @@
+
+A Coroutine is a piece of code that executes "in parallel", over a certain amount of time, or generally asynchronously, with another; in our particular case, it is code that should run for an arbitrarily long amount of frames, without getting in the way of the usual flow of your code.
+
+This **does not mean** that the Coroutine is operating on a separate thread, and thus can be written ignoring the usual constraints of the Update-Draw loop of the game engine; each step of the Coroutine should be able to be performed in one Update call, so don't go around looking for prime numbers in a Coroutine, chances are your game would stutter.
+
+# A simple example
+
+Let's say, for example, that you want to code a Component that, upon pressure of a key, will make your GameObject's move 200 units to the right: with only this requirement you will probably end up with something like this
+
+```csharp
+public void OnUpdate()
+{
+	if (DualityApp.Keyboard.KeyHit(Key.Space))
+	{
+		this.GameObj.Transform.Pos.MoveBy(new Vector2(200, 0));
+	}
+}
+```
+
+Which is fine, and works well for you.
+But now, let's assume you want the transition to be smooth; one way would be to move 1 unit per frame (let's ignore the time it would take or the fact that it would be dependent on the frame rate)
+
+```csharp
+private Vector2? movementTarget = null;
+private Vector2 movementDirection = Vector2.Zero;
+
+public void OnUpdate()
+{
+	if(movementTarget != null)
+	{
+		if((this.movementTarget.Value - this.GameObj.Transform.Pos.Xy).LengthSquared >= this.movementDirection.LengthSquared)
+		{
+			this.GameObj.Transform.Pos.MoveBy(movementDirection);
+		}
+		else
+		{
+			this.GameObj.Transform.Pos.MoveTo(this.movementTarget.Value);
+			this.movementTarget = this.Game;
+		}
+	}
+	else if (DualityApp.Keyboard.KeyHit(Key.Space))
+	{
+		this.movementTarget = this.GameObj.Transform.Pos.Xy + new Vector2(200, 0);
+		this.movementDirection = (this.movementTarget.Value - this.GameObj.Transform.Pos.Xy).Normalized;
+	}
+}
+```
+
+Again, this works, but the OnUpdate method is starting to appear cluttered with unnecessary code; wouldn't it be great if the part of code in charge of managing the movement was in another part?
+Sure, you could refactor the code to extract the method and have a `UpdateColor` or similar being called at the end or beginning of `OnUpdate`; good, but we can still do better. We can make a Coroutine out of it.
```

Comment:
Wait, why colors? So far it was only about movement?

Reply:
Oh, at start I had an example about changing color but was getting weird trying to find a more complex use case to use later in the example. Nice catch!